MP3 Tag Editor/SourceFileManager.cs

```diff
@@ -0,0 +1,112 @@
+        }
+        else if (Path.GetExtension(filePath) == ".xml")
+        {
+            return LoadXmlImportFile(filePath);
+        }
+
+        throw new InvalidSourcePathException(filePath, "The specified file is not a valid external source file.");
+    }
+
+    /// <summary>
+    /// Loads a CSV file of the metadata of a collection of MP3 files.
+    /// </summary>
+    /// <param name="filePath">The path of the file.</param>
+    /// <returns>A tuple containing a collection of properties and a 2D array of id information per song respectively</returns>
+    /// <exception cref="InvalidSourcePathException">Thrown when the specified CSV file does not exist.</exception>
+    private static (IEnumerable<string>, IEnumerable<string[]>) LoadCsvImportFile(string filePath)
+    {
+        //Assuming CSV is in format exported by mp3tag application
+        using (var csvParser = new TextFieldParser(filePath))
+        {
+            csvParser.SetDelimiters(";");
+
+            var idProperties = csvParser.ReadFields();
+            var idValues = new List<string[]>();
+            while (!csvParser.EndOfData)
+            {
+                idValues.Add(csvParser.ReadFields()?? new string[0]);
+            }
+            return (idProperties, idValues);
+        }
+    }
+
+    /// <summary>
+    /// Loads an XML file of the metadata for a collection of MP3 files.
+    /// </summary>
+    /// <param name="filePath">The path of the file</param>
+    /// <returns>An XmlDocument object of the given XML file</returns>
+    private static XmlDocument LoadXmlImportFile(string filePath)
+    {
+        var xmlFile = new XmlDocument();
+        xmlFile.Load(filePath);
+        return xmlFile;
+    }
+    private static void ModifyFromCsvTest()
+    {
+        var mp3DirTestPath = "ADD DIRPATH HERE";
+        var csvTestPath = "ADD FILEPATH HERE";
+
+        var mp3Files = Mp3FileManager.LoadMp3Files(mp3DirTestPath);
+        var csvFile = LoadImportFile(csvTestPath);
+
+        Mp3TagsManager.ModifyMp3Tags(mp3Files, Mp3Tag.Title, "Test");
+        Mp3FileManager.SaveMp3Files(mp3Files);
+
+        Console.WriteLine("Check");
+        Console.ReadLine();
+
+        Mp3TagsManager.ModifyMp3TagsWithCsv(mp3Files, csvFile.Item1, csvFile.Item2);
+        Mp3FileManager.SaveMp3Files(mp3Files);
+        Console.WriteLine("Done");
+        Console.ReadLine();
+    }
+
+    private static void ModifyFromXmlTest()
+    {
+        var mp3DirTestPath = "ADD DIRPATH HERE";
+        var xmlTestPath = "ADD FILEPATH HERE";
+
+        var mp3Files = Mp3FileManager.LoadMp3Files(mp3DirTestPath);
+        var csvFile = LoadImportFile(xmlTestPath);
+
+        Mp3TagsManager.ModifyMp3Tags(mp3Files, Mp3Tag.Title, "Test");
+        Mp3FileManager.SaveMp3Files(mp3Files);
+
+        Console.WriteLine("Check");
+        Console.ReadLine();
+
+        Mp3TagsManager.ModifyMp3TagsWithXml(mp3Files, csvFile);
+        Mp3FileManager.SaveMp3Files(mp3Files);
```

Comment:
This is a tricky one. I admit it's my fault. However, after some research, it looks like IEnumerable would not be the best type of parameter to use in this case. I have done some research and from what I can see we should use IReadOnlyCollection when possible and otherwise use ICollection instead of IEnumerable. You can look into it as well and we can skip this PR and add it as an issue/enhancement and update it in a separate branch because this change will affect a lot of functions.

You can look up 'Possible multiple enumeration' for reference.
Also, check this out: [https://stackoverflow.com/a/10113331](https://stackoverflow.com/a/10113331)

Reply:
This would definitely need a meeting

Reply:
I think just keep it as an issue, this project is already stretching past its deadline.